File: ttd-dr/src/llm_client.py
"""
LLM Client for TTD-DR
Supports OpenAI and Anthropic models
"""

import os
from typing import Optional, List, Dict
from abc import ABC, abstractmethod


class LLMClient(ABC):
    """Abstract base class for LLM clients"""

    @abstractmethod
    def generate(self, prompt: str, temperature: float = 0.7, max_tokens: int = 4096) -> str:
        """Generate a response from the LLM"""
        pass

    @abstractmethod
    def generate_multiple(self, prompt: str, n: int, temperature: float = 0.7, max_tokens: int = 4096) -> List[str]:
        """Generate multiple responses for self-evolution"""
        pass


class OpenAIClient(LLMClient):
    """OpenAI GPT client"""

    def __init__(self, model: str = "gpt-4", api_key: Optional[str] = None):
        try:
            from openai import OpenAI
        except ImportError:
            raise ImportError("Please install openai: pip install openai")

        self.client = OpenAI(api_key=api_key or os.getenv("OPENAI_API_KEY"))
        self.model = model

    def generate(self, prompt: str, temperature: float = 0.7, max_tokens: int = 4096) -> str:
        """Generate a single response"""
        response = self.client.chat.completions.create(
            model=self.model,
            messages=[{"role": "user", "content": prompt}],
            temperature=temperature,
            max_tokens=max_tokens
        )
        return response.choices[0].message.content

    def generate_multiple(self, prompt: str, n: int, temperature: float = 0.7, max_tokens: int = 4096) -> List[str]:
        """Generate multiple responses with varied parameters for exploration"""
        responses = []
        # Vary temperature for diversity (Section 2.2, Step 1: Initial States)
        temperatures = [temperature + (i * 0.1) for i in range(n)]

        for temp in temperatures[:n]:
            try:
                response = self.client.chat.completions.create(
                    model=self.model,
                    messages=[{"role": "user", "content": prompt}],
                    temperature=min(temp, 1.0),
                    max_tokens=max_tokens
                )
                responses.append(response.choices[0].message.content)
            except Exception as e:
                print(f"Error generating response: {e}")
                responses.append("")

        return responses


class AnthropicClient(LLMClient):
    """Anthropic Claude client"""

    def __init__(self, model: str = "claude-3-5-sonnet-20241022", api_key: Optional[str] = None):
        try:
            from anthropic import Anthropic
        except ImportError:
            raise ImportError("Please install anthropic: pip install anthropic")

        self.client = Anthropic(api_key=api_key or os.getenv("ANTHROPIC_API_KEY"))
        self.model = model

    def generate(self, prompt: str, temperature: float = 0.7, max_tokens: int = 4096) -> str:
        """Generate a single response"""
        response = self.client.messages.create(
            model=self.model,
            max_tokens=max_tokens,
            temperature=temperature,
            messages=[{"role": "user", "content": prompt}]
        )
        return response.content[0].text

    def generate_multiple(self, prompt: str, n: int, temperature: float = 0.7, max_tokens: int = 4096) -> List[str]:
        """Generate multiple responses with varied parameters"""
        responses = []
        temperatures = [temperature + (i * 0.1) for i in range(n)]

        for temp in temperatures[:n]:
            try:
                response = self.client.messages.create(
                    model=self.model,
                    max_tokens=max_tokens,
                    temperature=min(temp, 1.0),
                    messages=[{"role": "user", "content": prompt}]
                )
                responses.append(response.content[0].text)
            except Exception as e:
                print(f"Error generating response: {e}")
                responses.append("")

        return responses


def create_llm_client(provider: str, model: str, api_key: Optional[str] = None) -> LLMClient:
    """Factory function to create LLM client"""
    if provider.lower() == "openai":
        return OpenAIClient(model=model, api_key=api_key)
    elif provider.lower() == "anthropic":
        return AnthropicClient(model=model, api_key=api_key)
    else:
        raise ValueError(f"Unsupported provider: {provider}")

File: ttd-dr-ts/src/llm-client.ts
/**
 * LLM Client for TTD-DR
 * Supports OpenAI and Anthropic models
 */

import OpenAI from 'openai';
import Anthropic from '@anthropic-ai/sdk';

/**
 * Abstract base class for LLM clients
 */
export abstract class LLMClient {
  abstract generate(
    prompt: string,
    temperature?: number,
    maxTokens?: number
  ): Promise<string>;

  abstract generateMultiple(
    prompt: string,
    n: number,
    temperature?: number,
    maxTokens?: number
  ): Promise<string[]>;
}

/**
 * OpenAI GPT client
 */
export class OpenAIClient extends LLMClient {
  private client: OpenAI;
  private model: string;

  constructor(model: string = 'gpt-4', apiKey?: string) {
    super();
    this.client = new OpenAI({
      apiKey: apiKey || process.env.OPENAI_API_KEY,
    });
    this.model = model;
  }

  async generate(
    prompt: string,
    temperature: number = 0.7,
    maxTokens: number = 4096
  ): Promise<string> {
    const response = await this.client.chat.completions.create({
      model: this.model,
      messages: [{ role: 'user', content: prompt }],
      temperature,
      max_tokens: maxTokens,
    });

    return response.choices[0]?.message?.content || '';
  }

  async generateMultiple(
    prompt: string,
    n: number,
    temperature: number = 0.7,
    maxTokens: number = 4096
  ): Promise<string[]> {
    const responses: string[] = [];

    // Vary temperature for diversity (Section 2.2, Step 1: Initial States)
    const temperatures = Array.from({ length: n }, (_, i) => temperature + i * 0.1);

    for (let i = 0; i < n; i++) {
      try {
        const temp = Math.min(temperatures[i], 1.0);
        const response = await this.client.chat.completions.create({
          model: this.model,
          messages: [{ role: 'user', content: prompt }],
          temperature: temp,
          max_tokens: maxTokens,
        });

        responses.push(response.choices[0]?.message?.content || '');
      } catch (error) {
        console.error(`Error generating response ${i + 1}:`, error);
        responses.push('');
      }
    }

    return responses;
  }
}

/**
 * Anthropic Claude client
 */
export class AnthropicClient extends LLMClient {
  private client: Anthropic;
  private model: string;

  constructor(model: string = 'claude-3-5-sonnet-20241022', apiKey?: string) {
    super();
    this.client = new Anthropic({
      apiKey: apiKey || process.env.ANTHROPIC_API_KEY,
    });
    this.model = model;
  }

  async generate(
    prompt: string,
    temperature: number = 0.7,
    maxTokens: number = 4096
  ): Promise<string> {
    const response = await this.client.messages.create({
      model: this.model,
      max_tokens: maxTokens,
      temperature,
      messages: [{ role: 'user', content: prompt }],
    });

    const content = response.content[0];
    return content.type === 'text' ? content.text : '';
  }

  async generateMultiple(
    prompt: string,
    n: number,
    temperature: number = 0.7,
    maxTokens: number = 4096
  ): Promise<string[]> {
    const responses: string[] = [];
    const temperatures = Array.from({ length: n }, (_, i) => temperature + i * 0.1);

    for (let i = 0; i < n; i++) {
      try {
        const temp = Math.min(temperatures[i], 1.0);
        const response = await this.client.messages.create({
          model: this.model,
          max_tokens: maxTokens,
          temperature: temp,
          messages: [{ role: 'user', content: prompt }],
        });

        const content = response.content[0];
        responses.push(content.type === 'text' ? content.text : '');
      } catch (error) {
        console.error(`Error generating response ${i + 1}:`, error);
        responses.push('');
      }
    }

    return responses;
  }
}

/**
 * Factory function to create LLM client
 */
export function createLLMClient(
  provider: 'openai' | 'anthropic',
  model: string,
  apiKey?: string
): LLMClient {
  switch (provider.toLowerCase()) {
    case 'openai':
      return new OpenAIClient(model, apiKey);
    case 'anthropic':
      return new AnthropicClient(model, apiKey);
    default:
      throw new Error(`Unsupported provider: ${provider}`);
  }
}
